core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSailConnection.java

```diff
@@ -0,0 +1,1026 @@
+                return false;
+            }
+
+            @Override
+            public boolean add(Resource resource, IRI iri, Value value, Resource... resources) {
+                return false;
+            }
+
+            @Override
+            public boolean clear(Resource... resources) {
+                return false;
+            }
+
+            @Override
+            public boolean remove(Resource resource, IRI iri, Value value, Resource... resources) {
+                return false;
+            }
+
+            @Override
+            public Model filter(Resource resource, IRI iri, Value value, Resource... resources) {
+                return null;
+            }
+
+            @Override
+            public Set<Resource> subjects() {
+                return null;
+            }
+
+            @Override
+            public Set<IRI> predicates() {
+                return null;
+            }
+
+            @Override
+            public Set<Value> objects() {
+                return null;
+            }
+
+            @Override
+            public ValueFactory getValueFactory() {
+                return null;
+            }
+
+            @Override
+            public Iterator<Statement> match(Resource resource, IRI iri, Value value, Resource... resources) {
+                return null;
+            }
+
+            @Override
+            public int size() {
+                return 0;
+            }
+
+            @Override
+            public boolean isEmpty() {
+                return false;
+            }
+
+            @Override
+            public boolean contains(Object o) {
+                return false;
+            }
+
+            @Override
+            public Iterator<Statement> iterator() {
+                return null;
+            }
+
+            @Override
+            public Object[] toArray() {
+                return new Object[0];
+            }
+
+            @Override
+            public <T> T[] toArray(T[] a) {
+                return null;
+            }
+
+            @Override
+            public boolean add(Statement statement) {
+                return false;
+            }
+
+            @Override
+            public boolean remove(Object o) {
+                return false;
+            }
+
+            @Override
+            public boolean containsAll(Collection<?> c) {
+                return false;
+            }
+
+            @Override
+            public boolean addAll(Collection<? extends Statement> c) {
+                return false;
+            }
+
+            @Override
+            public boolean removeAll(Collection<?> c) {
+                return false;
+            }
+
+            @Override
+            public boolean retainAll(Collection<?> c) {
+                return false;
+            }
+
+            @Override
+            public void clear() {
+
+            }
+        };
+    }
+
+
+    protected int applyRules(Model model) throws SailException {
```

Comment:
Yeah, because the old reasoner applied all the rules in a stage right before commit it needed to keep the changes in memory until then. My approach doesn't require this, because all reasoning is done while the data is added.

There are two expections. 
1. A tbox statement is added.
2. A statement is removed.

Both of these will cause the reasoner to recalculate the inference tables and re run reasoning on all the triples in the store. I could make some benchmarks for this, because it causes the reasoner to slow down a lot, pretty much matching the current implementation.

If you can see any use for keeping a list of added and removed statements for my two exceptions then I would be very happy with some feedback.